CDAP-15985 MixPanel Batch Source plugin

YevheniiChekanskyi opened the pull request:
WIKI: https://wiki.cask.co/display/CE/Mixpanel+Batch+Source
JIRA: https://issues.cask.co/browse/CDAP-15985

YevheniiChekanskyi committed:
CDAP-15985 MixPanel Batch Source plugin

YevheniiChekanskyi committed:
CDAP-15985 MixPanel Batch Source plugin - review fixes

YevheniiChekanskyi commented:
@albertshau, @mikkin, @CuriousVini could you please review?

mikkin commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSourceConfig.java:
```diff
@@ -0,0 +1,192 @@
+    if (mixPanelUrl == null || mixPanelUrl.isEmpty()) {
+      return "https://data.mixpanel.com/api/2.0/export/";
+    } else {
+      return mixPanelUrl;
```
Recommend doing basic URL validation if a custom URL is entered.

mikkin reviewed (commented):
Looks good, overall. Added one comment around URL validation.

YevheniiChekanskyi committed:
CDAP-15985 MixPanel Batch Source plugin - basic url validation

YevheniiChekanskyi commented:
@mikkin could you please review?

mikkin commented:
LGTM but I have asked @CuriousVini to take another pass since I am shadowing her in this review.

CuriousVini commented on checkstyle.xml:
```diff
@@ -0,0 +1,387 @@
+    <property name="messageFormat" value="$1"/>
+  </module>
+
+</module>
```
nit: new line

YevheniiChekanskyi replied:
done

CuriousVini commented on suppressions.xml:
```diff
@@ -0,0 +1,30 @@
+  <suppress checks="JavadocStyle" files=".*[/\\]src[/\\](main|integration)[/\\]java[/\\].*" />
+  <suppress checks="JavadocStyle" files=".*[/\\]src[/\\].*[/\\]internal[/\\].*" />
+
+</suppressions>
```
nit: new line

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelSplit.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * A no-op split.
+ */
+public class MixPanelSplit extends InputSplit implements Writable {
```
rename it to NoOpMixPanelSplit.

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelRecordReader.java:
```diff
@@ -0,0 +1,158 @@
+ * RecordReader implementation, which reads events in json format from MixPanel api.
+ */
+public class MixPanelRecordReader extends RecordReader<NullWritable, Text> {
+  private static final Gson gson = new GsonBuilder().create();
```
static final are constants with all caps - `GSON`.

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSourceConfig.java:
```diff
@@ -0,0 +1,201 @@
+      new URL(getMixPanelUrl());
+    } catch (MalformedURLException e) {
+      failureCollector
+        .addFailure(String.format("Invalid URL '%s'", getMixPanelUrl()), "Change MixPanel url to valid")
```
all error messages and corrective actions should end with period.

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSourceConfig.java:
```diff
@@ -0,0 +1,201 @@
+  protected String toDate;
+
+  @Name(PROPERTY_EVENTS)
+  @Description("Comma separated list of events you would like to get data on.")
```
lets not use `you` in the description.

YevheniiChekanskyi replied:
removed

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSourceConfig.java:
```diff
@@ -0,0 +1,201 @@
+   */
+  public String getMixPanelUrl() {
+    if (mixPanelUrl == null || mixPanelUrl.isEmpty()) {
+      return "https://data.mixpanel.com/api/2.0/export/";
```
make it a string constant

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSourceConfig.java:
```diff
@@ -0,0 +1,201 @@
+  public String getMixPanelUrl() {
+    if (mixPanelUrl == null || mixPanelUrl.isEmpty()) {
+      return "https://data.mixpanel.com/api/2.0/export/";
+    } else {
```
there is no need for else here.

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSourceConfig.java:
```diff
@@ -0,0 +1,201 @@
+  public static final String PROPERTY_FILTER = "filter";
+  public static final String PROPERTY_URL = "mixPanelUrl";
+
+  private static final Pattern DATE_REGEX = Pattern.compile("\\d\\d\\d\\d-\\d\\d-\\d\\d");
```
can `^\d{4}-\d{2}-\d{2}$` be used instead?

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSource.java:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  private void validateConfiguration(FailureCollector failureCollector) {
+    IdUtils.validateReferenceName(config.referenceName, failureCollector);
```
this can be moved to `config.validate` method

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelBatchSource.java:
```diff
@@ -0,0 +1,87 @@
+  public static final String NAME = "MixPanel";
+  private static final Schema MIX_PANEL_RECORD_SCHEMA = Schema.recordOf(
+    "mixPanelRecord", Schema.Field.of("event", Schema.of(Schema.Type.STRING))
+  );
```
can be moved to previous line.

YevheniiChekanskyi replied:
done

CuriousVini commented on docs/MixPanel-batchsource.md:
```diff
@@ -0,0 +1,22 @@
+
+**Events:** Comma separated list of events you would like to get data on.
+
+**Filter:** Expression to filter events by.
```
what can be the value of a filter?

YevheniiChekanskyi replied:
Syntax is complex, added link to official documentation.

CuriousVini commented on docs/MixPanel-batchsource.md:
```diff
@@ -0,0 +1,22 @@
+
+**API Secret:** API secret used to authenticate in MixPanel API.
+
+**From date:** Start date for reports data.
```
add the expected format here.

YevheniiChekanskyi replied:
done

CuriousVini commented on docs/MixPanel-batchsource.md:
```diff
@@ -0,0 +1,22 @@
+
+**From date:** Start date for reports data.
+
+**To date:** End date for reports data.
```
same as above

YevheniiChekanskyi replied:
done

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelRecordReader.java:
```diff
@@ -0,0 +1,158 @@
+    HttpPost request = new HttpPost(config.getMixPanelUrl());
+    fillRequest(config, request);
+
+    response = httpClient.execute(request, context);
```
does this mean whole data is held in the response? Is there hadoop connector for mixpanel integration?

YevheniiChekanskyi replied:
Data retrieved by scanner sequentially from response stream, it is not grabbed completely after response object returned.

There is no connectors for this. Api is very simple itself - single HTTP query returning single JSONL file(each line of file is some json object representing event).

CuriousVini replied:
So basically response contents has all the events and scanner just scans through each line right? That means all the content will be held in memory?

CuriousVini commented on src/main/java/io/cdap/plugin/mixpanel/source/batch/MixPanelRecordReader.java:
```diff
@@ -0,0 +1,158 @@
+  private static final Gson gson = new GsonBuilder().create();
+  private String currentEvent;
+  private CloseableHttpClient httpClient;
+  private Scanner dataScanner;
```
scanner should be closed as well right?

YevheniiChekanskyi replied:
done

CuriousVini reviewed (changes requested):
Can you add an example of event json string read by the source? Is there any metadata associated with event such as timestamp that we should propagate as well?

YevheniiChekanskyi committed:
CDAP-15985 MixPanel Batch Source plugin - address review comments

YevheniiChekanskyi commented:
@CuriousVini thanks for review. I addressed all of your comments.
Regarding example of event, copied from documentation:
```
{
  "event": "Viewed report",
    "properties": {
        "distinct_id": "foo",
        "time": 1518314400,
        "$os": "Linux",
        "$browser": "Chrome",
        "Project ID": "3",
        "mp_country_code": "US"
    }
}
```
I think _time_ field will be always presented(but I am not sure for 100%) and we can extract it, but there was no such requirement in original requirements. Do you think we need propagate it? Desired time range specified in plugin parameters, also events most likely will be parsed and processed by pipeline. I don't think we need to introduce extra field here.

CuriousVini commented:
Left a comment here: https://wiki.cask.co/display/CE/Mixpanel+Batch+Source?focusedCommentId=22676513#comment-22676513
Lets put the implementation on-hold and add the details to the design doc.

YevheniiChekanskyi committed:
CDAP-15985 MixPanel Batch Source plugin - support schema fetching.

YevheniiChekanskyi committed:
CDAP-15985 MixPanel Batch Source plugin - refactoring.

YevheniiChekanskyi committed:
CDAP-15985 MixPanel Batch Source plugin - move configs around.

YevheniiChekanskyi commented:
@albertshau, @mikkin, @CuriousVini could you please review? There still open question about widgets naming/placement(see latest comments in https://wiki.cask.co/display/CE/Mixpanel+Batch+Source), but looks like Bhooshan Mogal, Prateek Duble and me found suitable design. I don't expect significant changes in code anymore.
